pkg/expansion/expand.go

```diff
@@ -0,0 +1,224 @@
+	emptyGVK := schema.GroupVersionKind{}
+	if resultantGVK == emptyGVK {
+		return nil, fmt.Errorf("cannot expand generator for template with empty generatedGVK")
+	}
+
+	src, ok, err := unstructured.NestedMap(generator.Object, sourcePath(srcPath)...)
+	if err != nil {
+		return nil, fmt.Errorf("could not extract source field from unstructured: %s", err)
+	}
+	if !ok {
+		return nil, fmt.Errorf("could not find source field %q in generator", srcPath)
+	}
+
+	return createResultantKind(resultantGVK, src, mutators)
+}
+
+func convertUnstructuredToTyped(u *unstructured.Unstructured, obj interface{}) error {
+	if u == nil {
+		return fmt.Errorf("cannot convert nil unstructured to type")
+	}
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.UnstructuredContent(), obj)
+	return err
+}
+
+func convertTemplateExpansion(u *unstructured.Unstructured) (mutationsunversioned.TemplateExpansion, error) {
+	te := mutationsunversioned.TemplateExpansion{}
+	err := convertUnstructuredToTyped(u, &te)
+	return te, err
+}
+
+func convertAssign(u *unstructured.Unstructured) (mutationsunversioned.Assign, error) {
+	a := mutationsunversioned.Assign{}
+	err := convertUnstructuredToTyped(u, &a)
+	return a, err
+}
+
+func convertAssignMetadata(u *unstructured.Unstructured) (mutationsunversioned.AssignMetadata, error) {
+	am := mutationsunversioned.AssignMetadata{}
+	err := convertUnstructuredToTyped(u, &am)
+	return am, err
+}
+
+func convertModifySet(u *unstructured.Unstructured) (mutationsunversioned.ModifySet, error) {
+	ms := mutationsunversioned.ModifySet{}
+	err := convertUnstructuredToTyped(u, &ms)
+	return ms, err
+}
+
+// sortResources sorts a list of resources into mutators, generators, template expansions and return
+// them respectively.
+func sortResources(resources []*unstructured.Unstructured) ([]*unstructured.Unstructured, []*unstructured.Unstructured, []*unstructured.Unstructured) {
+	var generators []*unstructured.Unstructured
+	var mutators []*unstructured.Unstructured
+	var templates []*unstructured.Unstructured
+
+	for _, r := range resources {
+		k := r.GetKind()
+		_, isMutator := MutatorTypes[k]
+		switch {
+		case isMutator:
+			mutators = append(mutators, r)
+		case k == "TemplateExpansion":
+			templates = append(templates, r)
+		default:
+			generators = append(generators, r)
+		}
+	}
+
+	return generators, mutators, templates
+}
+
+func sortMutators(mutators []types.Mutator) {
+	sort.SliceStable(mutators, func(x, y int) bool {
+		return mutatorSortKey(mutators[x]) < mutatorSortKey(mutators[y])
+	})
+}
+
+func mutatorSortKey(mutator types.Mutator) string {
+	return mutator.String()
+}
+
+func createResultantKind(gvk schema.GroupVersionKind, source map[string]interface{}, mutators []types.Mutator) (*unstructured.Unstructured, error) {
+	resource := unstructured.Unstructured{}
+	resource.SetUnstructuredContent(source)
+	resource.SetGroupVersionKind(gvk)
+	sortMutators(mutators)
+
+	for _, m := range mutators {
```

Comment:
We could call out to mutation system to do this.

The challenge is to create a `Match` test for `Source`. Somehow we will need to maintain metadata that tells the system the source of the object (annotations/labels should be avoided b/c then it may be possible to the user to trick the system by injecting bad data)